internal/core/src/storage/SafeQueue.h

```diff
@@ -0,0 +1,70 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma
+
+#include <mutex>
+#include <queue>
+#include <utility>
+
+namespace milvus {
+template <typename T>
+class SafeQueue {
+ public:
+    SafeQueue(void) {
+    }
+    ~SafeQueue() {
+    }
+
+    bool
+    empty() {
+        std::lock_guard<std::mutex> l(mutex_);
+        return queue_.empty();
+    }
+
+    void
+    size() {
+        std::lock_guard<std::mutex> l(mutex_);
+        return queue_.size();
+    }
+
+    // Add a element to queue
+    void
+    enqueue(T t) {
+        std::lock_guard<std::mutex> l(mutex_);
+        queue_.push(t);
+    }
+
+    // Get a available element from queue
+    bool
+    dequeue(T& t) {
+        std::lock_guard<std::mutex> l(mutex_);
+
+        if (queue_.empty()) {
+            return false;
+        }
+
+        t = std::move(queue_.front());
+        queue_.pop();
+
+        return true;
+    }
+
+ private:
+    std::queue<T> queue_;
+    mutable std::mutex mutex_;
```

Comment:
use shard_mutex will be better.
[[share_mutex](https://en.cppreference.com/w/cpp/thread/shared_mutex)](https://en.cppreference.com/w/cpp/thread/shared_mutex)